lib/statsd.js

```diff
@@ -242,22 +248,41 @@ Client.prototype.sendStat = function (stat, value, type, sampleRate, tags, callb
  * @param callback {Function=} Callback when message is done being delivered (only if maxBufferSize == 0). Optional.
  */
 Client.prototype.send = function (message, tags, callback) {
-  let mergedTags = this.globalTags;
-  if (tags && typeof tags === 'object') {
-    mergedTags = helpers.overrideTags(mergedTags, tags, this.telegraf);
-  }
-  if (mergedTags.length > 0) {
+  const tagsString = getTagString(this.globalTags, this.globalTagsMemo, tags, this.telegraf, this.tagSeparator);
+  if (tagsString.length > 0) {
     if (this.telegraf) {
       message = message.split(':');
-      message = `${message[0]},${mergedTags.join(',').replace(/:/g, '=')}:${message.slice(1).join(':')}`;
+      message = `${message[0]},${tagsString.replace(/:/g, '=')}:${message.slice(1).join(':')}`;
     } else {
-      message += `|${this.tagPrefix}${mergedTags.join(this.tagSeparator)}`;
+      message += `|${this.tagPrefix}${tagsString}`;
     }
   }
 
   this._send(message, callback);
 };
 
+// eslint-disable-next-line require-jsdoc
+function getTagString(globalTags, globalTagsMemo, tags, telegraf, tagSeparator) {
+  if (!tags) {
+    return globalTagsMemo;
+  }
+  const isTagsArray = Array.isArray(tags);
+  if (isTagsArray) {
+    if (tags.length === 0) {
+      return globalTagsMemo;
+    }
+  }
+  else if (Object.keys(tags).length === 0) {
+    return globalTagsMemo;
+  }
+  if (isTagsArray || Array.isArray(globalTags)) {
```

Comment:
the lack of typing here (not really your fault) makes it really hard to follow what the assumptions here are... to is it possible for `tags` to be an array but for `globalTags` to be an object? or the other way around?

does `overrideTagsToString` work if you pass in objects for `globalTags` or `tags`?

Reply:
tags could be array or object. globalTags could be array or object. Welcome to the wild west.

`overrideTagsToString` does work for any combination of objects/arrays. 

`overrideTags2` only works if globalTags is an object, and also tags is an object

Reply:
btw, the unit tests seem pretty good, but i was thinking i'd also test this by testing on a manual image deployment of xover in a small staging region